specs/core/0_beacon-chain.md

```diff
@@ -1663,7 +1695,7 @@ def process_ejections(state: BeaconState) -> None:
 If the following are satisfied:
 
 * `state.finalized_slot > state.validator_registry_latest_change_slot`
-* `state.latest_crosslinks[shard].slot > state.validator_registry_latest_change_slot` for every shard number `shard` in `shard_committee` from `shard_committee_at_slot` in `state.shard_committees_at_slots`
+* `state.latest_crosslinks[shard].slot > state.validator_registry_latest_change_slot` for every shard number `shard` in `[(state.current_epoch_start_shard + i) % SHARD_COUNT for i in range(get_current_epoch_committees_per_slot(state) * EPOCH_LENGTH)]` (that is, for every shard in the current committees)
```

Comment:
Should this actually be conditioned upon the previous epoch crosslink shards rather than the current? It is nearly impossible (100% impossible if we move epoch transition to last slot in epoch) that all of the crosslinks from the first epoch of a new shuffling be crosslinked during that epoch.

Neither prev or current are that clean. Both concede that this can't be satisfied within 1 epoch.

Reply:
I might be missing something here.

Say we are at the epoch transition at slot 128 (the end of epoch 64-127). The current committees are the committees related to slots 64-127. The committees at slot 127 have had 1 (or 0 if we do epoch transition at the end of 127) slots to be included. So the committees at slot 127 will not be able to crosslink their shard by this state transition (especially considering `MIN_ATTESTATION_INCLUSION_DELAY`) and will only have potentially successfully crosslinked by the end of the next epoch.

Reply:
Aah, I see, yes you're right; I was thinking of justification.

So in this case regardless of how we do it, shards would need to switch over only once every two epochs. I do think it has to be `current` though, because if we used `previous` then we could have a situation where if the previous shards are eg. 0-31 and the current shards are 32-63, and the previous shards are crosslinked but the current shards are not then the current shards increment to 64-95, which means it skips over 32-63 without crosslinking them.

Reply:
agreed. previous isn't the solution.
This is a nasty snag. I'm not immediately seeing a good solution...

One potential would be to just have committees crosslink in the first 4/5 or something of the epoch with the remaining 1/5th is just attestation inclusion and finality gadget with shard as `2**64-1` and `shard_block_hash` as `0x0` or something.

Reply:
Leave as is for this PR and make note in an issue?

Reply:
Actually, it's not technically an efficiency decrease in the current code, because it turns out _both_ of the conditions for validator set rotation + start shard rotation require 2+ epochs to work. The conditions are:

If the following are satisfied:

* `state.finalized_slot > state.validator_registry_latest_change_slot`
* `state.latest_crosslinks[shard].slot > state.validator_registry_latest_change_slot` for every shard number shard in `[(state.current_epoch_start_shard + i) % SHARD_COUNT for i in range(get_current_epoch_committees_per_slot(state) * EPOCH_LENGTH)]` (that is, for every shard in the current committees)

The second rule takes two epochs because of what we discussed above. The first rule also takes two epochs because, well, finality takes two epochs.

Reply:
You can continually finalized epoch after epoch after epoch when the chain is running smoothly. This crosslink delay does not get better due to the shuffling remaining stagnant

Reply:
Yes, the finalized block can increase every epoch, but `state.finalized_slot > state.validator_registry_latest_change_slot` is not going to happen more than once every two blocks, because at time T you're finalizing at best T - 2 * EPOCH_LENGTH.

Reply:
oh, well we don't technically need the `validator_registry_latest_change_slot` to lag like that. We can easily have it be `validator_registry_latest_finalized_change_slot` or something and set it to the `finalized_slot` when the registry change happens. so the condition becomes "have we finalized something new since the last registry change" instead of "have we finalized a slot past the slot when the last registry change happened"

If you do that then the crosslinks become the holdup.

Reply:
Maybe some weirdness falls out here because the new finality could be based upon attestations included from the previous shuffling, but I don't think this will reduce the fault tolerance by more than `MAX_BALANCE_CHURN_QUOTIENT`

Reply:
Taking this convo to #436